pkg/releaselock/releaselock.go

```diff
@@ -0,0 +1,89 @@
+package releaselock
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"net/rpc"
+	"strconv"
+
+	"github.com/prashanthpai/sunrpc"
+	"k8s.io/klog/v2"
+)
+
+const (
+	pmapProgramNumber  = uint32(100021)
+	pmapProgramVersion = uint32(4)
+	programNumber      = uint32(200002)
+	programVersion     = uint32(1)
+	procedureNumber    = uint32(1)
+	procedureName      = "IN_BAND_PROPRIETARY_LOCK_OPS_PROG.IN_BAND_PROPRIETARY_LOCK_OPS_V1"
+	rpcPort            = "111"
+	protocol           = "tcp"
+)
+
+type releaseLockResponse struct {
+	status releaseLockStatus
+}
+
+type releaseLockStatus uint
+
+// ReleaseLock calls the NFS server to remove all advisory locks on the given GKE node.
+// hostIP is the internal IP address of the Filestore instance.
+// clientIP is the internal IP address of the GKE node.
+func ReleaseLock(hostIP, clientIP string) error {
+	procedureID := sunrpc.ProcedureID{
+		ProgramNumber:   programNumber,
+		ProgramVersion:  programVersion,
+		ProcedureNumber: procedureNumber,
+	}
+	procedure := sunrpc.Procedure{
+		ID:   procedureID,
+		Name: procedureName,
+	}
+	if err := sunrpc.RegisterProcedure(procedure, true /* validateProcName */); err != nil {
+		return fmt.Errorf("failed to register procedure %+v: %q", procedure, err)
+	}
+
+	// Get port from portmapper.
+	hostAddress := fmt.Sprintf("%s:%s", hostIP, rpcPort)
+	port, err := sunrpc.PmapGetPort(hostAddress, pmapProgramNumber, pmapProgramVersion, sunrpc.IPProtoTCP)
+	if err != nil {
+		return fmt.Errorf("failed to get port for %s: %q", hostAddress, err)
+	}
+
+	// Connect to RPC server.
+	serverAddress := fmt.Sprintf("%s:%s", hostIP, strconv.Itoa(int(port)))
+	conn, err := net.Dial(protocol, serverAddress)
+	if err != nil {
+		return fmt.Errorf("failed to connect to RPC server at address %s: %q", serverAddress, err)
+	}
+
+	// Get notified when server closes the connection.
+	notifyClose := make(chan io.ReadWriteCloser)
+	go func() {
+		for rwc := range notifyClose {
+			conn := rwc.(net.Conn)
+			klog.Infof("Server %s disconnected.", conn.RemoteAddr().String())
+		}
+	}()
+
+	// Create client using sunrpc codec.
+	client := rpc.NewClientWithCodec(sunrpc.NewClientCodec(conn, notifyClose))
+
+	klog.Infof("Calling NFS Server %q to release all locks on node %q.", serverAddress, clientIP)
+
+	var response releaseLockResponse
+	ipBinary := binary.BigEndian.Uint32(net.ParseIP(clientIP).To4())
```

Comment:
We should check for the invalid IP case. i.e return an error if net.ParseIP(clientIP).To4() is nil